narwhal/types/src/tests/error_test.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use prometheus::IntGauge;
+
+use super::DagError;
+use crate::metered_channel::{channel, Receiver, Sender, WithPermit};
+use std::time::Duration;
+
+pub struct Processor {
+    input: Receiver<Result<usize, DagError>>,
+    output: Sender<usize>,
+}
+
+impl Processor {
+    pub fn new(input: Receiver<Result<usize, DagError>>, output: Sender<usize>) -> Self {
+        Self { input, output }
+    }
+
+    pub fn spawn(input: Receiver<Result<usize, DagError>>, output: Sender<usize>) {
+        tokio::spawn(async move {
+            let mut processor = Processor::new(input, output);
+            processor.run().await;
+        });
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                Some((permit, Some(res_value))) = self.output.with_permit(self.input.recv())  => {
+                    permit.send(res_value.unwrap());
+                }
```

Comment:
I am thinking whether we could better prove the case of using the `with_permit` by having at least two branches in this loop. 
* One would just consume from the `input` channel and probably add to a secondary one (maybe to a FuturesUnordered list or something)
* And one which would try to take the next element from the list and acquire the permit

That would reveal/test the case when a branch (the one that uses the `with_permit`) gets cancelled because another one is ready to execute (since the `output` channel is full and we stuck on waiting for permit).

Ex
```
let waiting = FuturesUnordered::new();

loop {
    tokio::select! {
         Some(element) = self.input.recv() => {
               waiting.push(async move {
                    element
               });
        }
        Some((permit, Some(res_value))) = self.output.with_permit(waiting.next())  => {
             permit.send(res_value.unwrap());
         }
     }
}
```

@huitseeker what do you think?

Reply:
Yes, you are correct @akichidis . Testing this requires testing with a future that's explicitly not cancel-safe, and in fact `mychannel.recv()` is cancel-safe, so we're not testing anything in the current state of things. I've added a commit that fixes this, thanks for the feedback.